### ISSUE TITLE
Assert-PSRule hangs or throws an error when using rules generated from policy

### ISSUE THREAD

**Issue by VeraBE** [Azure/PSRule.Rules.Azure#1876]
**Description of the issue**

I'm trying to run the rules generated with `Export-AzPolicyAssignmentRuleData`, using `Assert-PSRule`, but the command either throws an error or hangs when processing the new rules (right after `VERBOSE: [PSRule][D] -- Discovering rules in: path\definitions-export-c4d3ac5d.Rule.jsonc`; I left it running for over 8 hours and nothing changed)

**To Reproduce**

Steps to reproduce the issue:

Run the following command:
```
Assert-PSRule -Module "PSRule.Rules.Azure" -Path "rulesFromPolicies" -InputPath "C:\Users\verabo.REDMOND\OneDrive - Microsoft\Documents\Testing" -Format File -OutputPath "analysisResults.json" -OutputFormat Sarif -Verbose
```

In the folder "rulesFromPolicies" there's only one file, with this content:
```
[
   {
      // Synopsis: "Audit enabling of resource logs. This enables you to recreate activity trails to use for investigation purposes; when a security incident occurs or when your network is compromised"
      "apiVersion":"github.com/microsoft/PSRule/v1",
      "kind":"Rule",
      "metadata":{
         "name":"Azure.Policy.c397c7e52a57",
         "ref":"Azure.Policy.c397c7e52a57",
         "tags":{
            "Azure.Policy/category":"Internet of Things"
         },
         "annotations":{
            "Azure.Policy/id":"/providers/Microsoft.Authorization/policyDefinitions/383856f8-de7f-44a2-81fc-e5135b5c2aa4",
            "Azure.Policy/version":"3.0.1"
         }
      },
      "spec":{
         "type":[
            "Microsoft.Devices/IotHubs"
         ],
         "where":{
            "equals":"Microsoft.Devices/IotHubs",
            "type":"."
         },
         "condition":{
            "field":"resources",
            "allOf":[
               {
                  "greaterOrEquals":1,
                  "field":"properties.logs[? || (@.retentionPolicy.enabled == 'true' && (@.retentionPolicy.days == '0' || @.retentionPolicy.days >= '365') && @.enabled == 'true') || (@.enabled == 'true')]"
               }
            ],
            "where":{
               "type":".",
               "equals":"Microsoft.Insights/diagnosticSettings"
            }
         }
      }
   }
]
```

I added this to the ps-rule.yaml file in the repository, and also to a file with that name in the folder from where I'm executing (I wasn't sure where it should go), but it didn't make a difference: https://azure.github.io/PSRule.Rules.Azure/customization/enforce-custom-tags/#binding-type

If, instead, I run this other command: 
```
Assert-PSRule -Module "PSRule.Rules.Azure" -Path 'C:\Users\verabo.REDMOND\Repos\PSRule.Rules.Azure\.ps-rule' -InputPath "C:\Users\verabo.REDMOND\OneDrive - Microsoft\Documents\Testing" -Format File -OutputPath "analysisResults.json" -OutputFormat Sarif -Verbose
```
And I add the same file to that .ps-rule folder then it throws an error:

```
[...]
VERBOSE: [PSRule][D] -- Discovering rules in: C:\Users\verabo.REDMOND\Repos\PSRule.Rules.Azure\out\modules\PSRule.Rules.Azure\rules\Common.Selector.Rule.yaml
VERBOSE: [PSRule][D] -- Discovering rules in: C:\Users\verabo.REDMOND\Repos\PSRule.Rules.Azure\out\modules\PSRule.Rules.Azure\rules\Config.Rule.yaml
VERBOSE: [PSRule][D] -- Discovering rules in: C:\Users\verabo.REDMOND\Repos\PSRule.Rules.Azure\.ps-rule\definitions-export-c4d3ac5d.Rule.jsonc
An item with the same key has already been added.
At C:\Users\verabo.REDMOND\OneDrive - Microsoft\Documents\WindowsPowerShell\Modules\PSRule\2.6.0\PSRule.psm1:609 char:13
+             throw $_.Exception.GetBaseException();
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (:) [], ArgumentException
    + FullyQualifiedErrorId : An item with the same key has already been added.
```

From this part of the docs: https://azure.github.io/PSRule.Rules.Azure/customization/enforce-custom-tags/#testing-locally I understood that the Path option can be used to specify the path of the new rules, is that correct? Is it hanging in the first case because I am using it wrong?

**Module in use and version:**

- Module: PSRule.Rules.Azure
- Version: I ran the version in the main branch up to the latest commit now (https://github.com/Azure/PSRule.Rules.Azure/commit/ef1570a25c6b1850d0f2a9116fa23aaaf13b9157)

Captured output from `$PSVersionTable`:

```text
Name                           Value
----                           -----
PSVersion                      5.1.22621.608
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.22621.608
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

**Comment by BernieWhite** [comment]
@VeraBE PSRule will automatically find the `.ps-rule/` from the current working directory or you can set `-Path` to an alternative path as you have done. You usage seems fine to me.

It will recursively scan the path from `-InputPath` and `-Path`, maybe it was waiting for OneDrive to sync files and got stuck? It is reproducible between runs? and do you get the same issue working in a non-synchronized directory. I tried to do something similar on a local path and I got your `An item with the same key has already been added.` exception.

**Comment by BernieWhite** [comment]
Looks like `An item with the same key has already been added.` is an upstream issue in PSRule where `ref` and `name` are duplicate keys. We should fix it in PSRule but it's probably unnecessary to output both `ref` and `name` with the same identifier.

**Comment by VeraBE** [comment]
Is PR #1893 supposed to fix the "An item with the same key has already been added." error? I pulled the latest version from main, built it, added the rule and config back to ps-rule.yml and .ps-rule, imported the built module, ran `Assert-PSRule -Module "PSRule.Rules.Azure" -Path 'C:\Users\verabo.REDMOND\Repos\PSRule.Rules.Azure\.ps-rule' -InputPath "C:\Users\verabo.REDMOND\OneDrive - Microsoft\Documents\Testing" -Format File -OutputPath "analysisResults.json" -OutputFormat Sarif -Verbose` and I still see the same issue ;/

**Comment by VeraBE** [comment]
@BernieWhite Is PR #1893 supposed to fix the error "An item with the same key has already been added."? I pulled the latest version from main, built it, added back the new rule and config to ps-rule.yaml and .ps-rule, imported the built module, ran `Assert-PSRule -Module "PSRule.Rules.Azure" -Path 'C:\Users\verabo.REDMOND\Repos\PSRule.Rules.Azure\.ps-rule' -InputPath "C:\Users\verabo.REDMOND\OneDrive - Microsoft\Documents\Testing" -Format File -OutputPath "analysisResults.json" -OutputFormat Sarif -Verbose` and I still get the same error :/ Am I doing something wrong?

**Comment by BernieWhite** [comment]
@VeraBE #1893 should fix `Export-AzPolicyAssignmentRuleData`.

When `Export-AzPolicyAssignmentRuleData` is called to generate rules in the future is will only generate a `metadata.name` property for each rule instead of `metadata.name` and `metadata.ref` with the same value. The issue is caused by these values being set to the same value.

Upstream microsoft/PSRule#1354 will fix the issue if `metadata.name` and `metadata.ref` already exist and are the same value.

--- 

Also separately there is an issue that generates the same exception message that is fixed by #1894 (#1871) which is related to assignments using the same parameter name.

---

Hope that makes sense. But if you're testing and that doesn't work let me know.

**Comment by BernieWhite** [comment]
@VeraBE This one should be fixed in PSRule v2.7.0-B0049 onwards based on microsoft/PSRule#1370.
